### PR TITLE
[CHK-6732] Add refund support to update payment rsa request

### PR DESCRIPTION
**Pull request by kfalkowski11**
#### Overview

Previously, Bold-hosted/platform connector set ups were only able to process refunds through the platform. Adding support to be able to initialize refunds from both the platform and from Bold Checkout.

Linked platform connector change: https://github.com/bold-commerce/magento-1-platform-connector/pull/36

Resolves [CHK-6732](https://boldapps.atlassian.net/browse/CHK-6732)

[CHK-6732]: https://boldapps.atlassian.net/browse/CHK-6732?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Checkout/Api/Platform/OrderPayments.php [hunk 1]
```diff
@@ -20,7 +20,33 @@ public static function update(
         $payload = json_decode($request->getRawBody());
         /** @var Mage_Sales_Model_Order $order */
         $order = Mage::getModel('sales/order')->load($payload->payment->payment->parent_id);
-        if (self::isDelayedCapture($order) || $order->hasInvoices()) {
+
+        if($payload->payment->transaction->txn_type === 'refund') {
```
Comment (JosephLeedy) [inline Checkout/Api/Platform/OrderPayments.php:24]: There should be a space between the `if` statement and the condition.

File: Checkout/Api/Platform/OrderPayments.php [hunk 2]
```diff
@@ -56,6 +82,8 @@ public static function update(
         );
     }
 
+
+
```
Comment (JosephLeedy) [inline Checkout/Api/Platform/OrderPayments.php:86]: There are two extra line breaks here.

File: Checkout/Api/Platform/OrderPayments.php [hunk 3]
```diff
@@ -81,6 +109,33 @@ private static function invoice(
...
+     *
+     * @param Mage_Sales_Model_Order $order
+     * @param Mage_Sales_Model_Order_Invoice $invoice
+     * @return void
+     * @throws Exception
+     */
+    private static function creditmemo(
+        Mage_Sales_Model_Order $order, 
+        Mage_Sales_Model_Order_Invoice $invoice,
+        string $transactionId
```
Comment (JosephLeedy) [inline Checkout/Api/Platform/OrderPayments.php:123]: Specifying scalar method parameter types is not supported in PHP 5.6.

File: Checkout/Service/PaymentMethod.php [hunk 4]
```diff
@@ -200,16 +200,20 @@ public function refund(Varien_Object $payment, $amount)
...
         }
-        $orderGrandTotal = Mage::app()->getStore()->roundPrice($order->getGrandTotal());
-        $amount = Mage::app()->getStore()->roundPrice($amount);
-        if ($orderGrandTotal <= $amount) {
-            $transactionId = Bold_Checkout_Api_Bold_Payment::refundFull($order);
-            $payment->setTransactionId($transactionId)
-                ->setIsTransactionClosed(1)
-                ->setShouldCloseParentTransaction(true);
-            return $this;
+        if(null !== $payment->getCreditmemo()->getTransactionId()){
```
Comment (JosephLeedy) [inline Checkout/Service/PaymentMethod.php:203]: There should be a space between the `if` statement and the condition as well as between the condition and the opening curly brace.

Yoda conditions should not be used for clarity.

**Review by JosephLeedy — CHANGES_REQUESTED** [review]

**Review by jdart — APPROVED** [review]

**Review by JosephLeedy — APPROVED** [review]

**Review by TanyaP-sdet — APPROVED** [review]

File: Checkout/Api/Platform/OrderPayments.php [hunk 5]
```diff
@@ -20,7 +20,33 @@ public static function update(
...
+            $invoice = $invoiceCollection->getFirstItem();
+
+            try {
+                self::creditmemo($order, $invoice, $payload->payment->transaction->txn_id);
+            } catch (Exception $e) {
+                $error = new stdClass();
+                $error->message = $e->getMessage();
+                $error->code = 500;
+                $error->type = 'server.internal_error';
+                return Bold_Checkout_Rest::buildResponse($response, json_encode(['errors' => [$error]]));
```
Comment (jeremywiebeofbold) [inline Checkout/Api/Platform/OrderPayments.php:35]: If we have bugsnag connected to this project, this seems like a good place to log the exception.

Comment (jeremywiebeofbold) [reply Checkout/Api/Platform/OrderPayments.php:35]: Oh, this is the module. What about writing the exception to the logger, then?

Comment (kfalkowski11) [reply Checkout/Api/Platform/OrderPayments.php:36]: yep, added!

**Review by bold-eric-d — APPROVED** [review]

**Review by jeremywiebeofbold — APPROVED** [review]

**Review by JosephLeedy — APPROVED** [review]

**Pull request merged**